File: app/src/main/java/com/tydeya/familycircle/ui/livepart/memberpersonpage/details/MemberPersonFragment.java
package com.tydeya.familycircle.ui.livepart.memberpersonpage.details;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.MenuInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.PopupMenu;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.Toolbar;
import androidx.fragment.app.Fragment;
import androidx.navigation.NavController;
import androidx.navigation.fragment.NavHostFragment;

import com.bumptech.glide.Glide;
import com.tydeya.familycircle.App;
import com.tydeya.familycircle.R;
import com.tydeya.familycircle.data.familymember.FamilyMember;
import com.tydeya.familycircle.data.familymember.dto.FamilyMemberDto;
import com.tydeya.familycircle.domain.familyassistant.abstraction.FamilyAssistant;
import com.tydeya.familycircle.domain.familyassistant.details.FamilyAssistantImpl;
import com.tydeya.familycircle.domain.familyinteractor.abstraction.FamilyInteractorCallback;
import com.tydeya.familycircle.domain.familyinteractor.details.FamilyInteractor;
import com.tydeya.familycircle.domain.onlinemanager.details.OnlineInteractorImpl;
import com.tydeya.familycircle.ui.livepart.memberpersonpage.abstraction.MemberPersonPresenter;
import com.tydeya.familycircle.ui.livepart.memberpersonpage.abstraction.MemberPersonView;

import javax.inject.Inject;

import static com.tydeya.familycircle.utils.DipKt.getDp;


public class MemberPersonFragment extends Fragment implements MemberPersonView, FamilyInteractorCallback {

    private ImageView profileImage;
    private TextView nameText;
    private TextView onlineStatusText;
    private TextView birthdateText;
    private TextView zodiacSignText;
    private TextView workPlaceText;
    private TextView studyPlaceText;

    private Toolbar toolbar;
    private ImageButton settingsButton;

    private NavController navController;

    @Inject
    FamilyInteractor familyInteractor;

    @Inject
    OnlineInteractorImpl onlineInteractor;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        App.getComponent().injectFragment(this);

        View root = inflater.inflate(R.layout.fragment_family_member_view, container, false);

        toolbar = root.findViewById(R.id.family_member_view_toolbar);
        settingsButton = root.findViewById(R.id.family_member_view_settings);

        nameText = root.findViewById(R.id.family_member_view_name_text);
        profileImage = root.findViewById(R.id.family_view_photo);
        onlineStatusText = root.findViewById(R.id.family_member_view_online_text);
        birthdateText = root.findViewById(R.id.family_member_view_birthdate_text);
        zodiacSignText = root.findViewById(R.id.family_member_view_zodiac_sign);
        studyPlaceText = root.findViewById(R.id.family_member_view_study_place);
        workPlaceText = root.findViewById(R.id.family_member_view_work_place);

        return root;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        MemberPersonPresenter presenter = new MemberPersonPresenterImpl(this, getFamilyMember());

        toolbar.setNavigationOnClickListener(v -> {
            NavController navController = NavHostFragment.findNavController(this);
            navController.popBackStack();
        });

        navController = NavHostFragment.findNavController(this);
    }

    private FamilyMember getFamilyMember() {
        FamilyAssistant familyAssistant = new FamilyAssistantImpl(familyInteractor.getActualFamily());
        return familyAssistant.getUserByPhone(getArguments().getString("personFullPhoneNumber", ""));
    }

    @Override
    public void setCurrentData(FamilyMemberDto dto) {
        assert getContext() != null;
        nameText.setText(dto.getName());

        if (!dto.getImageAddress().equals("")) {
            profileImage.setPadding(0,0,0,0);
            Glide.with(getContext()).load(dto.getImageAddress()).into(profileImage);
        } else {
            int dpForPadding = getDp(getContext(), 20);
            profileImage.setPadding(dpForPadding,dpForPadding,dpForPadding,dpForPadding);
        }

        if (dto.getBirthDate().equals("")) {
            birthdateText.setText(getResources()
                    .getString(R.string.family_member_view_date_of_birth_not_known));

            zodiacSignText.setVisibility(View.GONE);
        } else {
            birthdateText.setText(dto.getBirthDate());
            zodiacSignText.setText(dto.getZodiacSign());
            zodiacSignText.setVisibility(View.VISIBLE);
        }

        if (dto.getStudyPlace().equals("")) {
            studyPlaceText.setVisibility(View.GONE);
        } else {
            studyPlaceText.setVisibility(View.VISIBLE);
            studyPlaceText.setText(dto.getStudyPlace());
        }

        if (dto.getWorkPlace().equals("")) {
            workPlaceText.setVisibility(View.GONE);
        } else {
            workPlaceText.setVisibility(View.VISIBLE);
            workPlaceText.setText(dto.getWorkPlace());
        }

        if (onlineInteractor.isUserOnline(dto.getPhone())) {
            onlineStatusText.setText(getContext().getString(R.string.online_text));
            onlineStatusText.setBackgroundColor(getResources().getColor(R.color.colorOnlineGreen));
        } else {
            onlineStatusText.setText(getContext().getString(R.string.offline_text));
            onlineStatusText.setBackgroundColor(getResources().getColor(R.color.colorGray));
        }

    }

    @Override
    public void setManagerMode(boolean managerMode) {
        if (managerMode) {
            settingsButton.setVisibility(View.VISIBLE);
            settingsButton.setEnabled(true);
            settingsButton.setOnClickListener(this::showPopUpSettingsMenu);
        } else {
            settingsButton.setVisibility(View.INVISIBLE);
            settingsButton.setEnabled(false);
        }
    }

    private void showPopUpSettingsMenu(View v) {
        PopupMenu popupMenu = new PopupMenu(getContext(), v);
        MenuInflater menuInflater = popupMenu.getMenuInflater();
        menuInflater.inflate(R.menu.settings_person_page_menu, popupMenu.getMenu());
        popupMenu.setOnMenuItemClickListener(item -> {
            switch (item.getItemId()) {
                case R.id.edit_person_page:
                    Bundle bundle = new Bundle();
                    bundle.putString("personFullPhoneNumber", getFamilyMember().getFullPhoneNumber());
                    navController.navigate(R.id.memberPersonEditFragment);
                    return true;
                case R.id.add_post_to_person_page:
                    return true;
                default:
                    return false;
            }
        });
        popupMenu.show();
    }

    @Override
    public void onResume() {
        super.onResume();
        familyInteractor.subscribe(this);
    }

    @Override
    public void memberDataUpdated() {
        setCurrentData(new FamilyMemberDto(getFamilyMember()));
    }

    @Override
    public void onPause() {
        super.onPause();
        familyInteractor.unsubscribe(this);
    }
}

File: app/src/main/java/com/tydeya/familycircle/presentation/ui/livepart/memberpersonpage/MemberPersonFragment.kt
package com.tydeya.familycircle.presentation.ui.livepart.memberpersonpage

import android.os.Bundle
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.PopupMenu
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import androidx.navigation.fragment.NavHostFragment
import androidx.recyclerview.widget.LinearLayoutManager
import com.bumptech.glide.Glide
import com.tydeya.familycircle.R
import com.tydeya.familycircle.data.constants.NavigateConsts
import com.tydeya.familycircle.data.familymember.FamilyMemberDto
import com.tydeya.familycircle.databinding.FragmentFamilyMemberViewBinding
import com.tydeya.familycircle.presentation.ui.livepart.memberpersonpage.twitterrecycler.TwitterRecyclerViewAdapter
import com.tydeya.familycircle.presentation.viewmodel.familyviewmodel.FamilyViewModel
import com.tydeya.familycircle.presentation.viewmodel.familyviewmodel.FamilyViewModelFactory
import com.tydeya.familycircle.utils.Resource
import com.tydeya.familycircle.utils.extensions.currentFamilyId
import com.tydeya.familycircle.utils.extensions.getUserPhone
import com.tydeya.familycircle.utils.extensions.popBackStack
import com.tydeya.familycircle.utils.extensions.toArrayList
import com.tydeya.familycircle.utils.getDp

class MemberPersonFragment : Fragment() {

    private var _binding: FragmentFamilyMemberViewBinding? = null
    private val binding get() = _binding!!

    private lateinit var familyViewModel: FamilyViewModel

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        _binding = FragmentFamilyMemberViewBinding.inflate(inflater, container, false)
        familyViewModel = ViewModelProviders
                .of(requireActivity(), FamilyViewModelFactory(requireActivity().currentFamilyId))
                .get(FamilyViewModel::class.java)

        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        requireArguments().getString(NavigateConsts.BUNDLE_FULL_PHONE_NUMBER)?.let {
            initFamilyView(it)
            initTweetRibbon(it)
        } ?: popBackStack()

        binding.toolbar.setNavigationOnClickListener {
            popBackStack()
        }
        binding.familyMemberViewAddTweet.setOnClickListener {
            NavHostFragment.findNavController(this).navigate(R.id.addTweetFragment)
        }
    }

    private fun initTweetRibbon(phone: String) {
        val twitterAdapter = TwitterRecyclerViewAdapter()
        binding.twitterRecycler.adapter = twitterAdapter
        binding.twitterRecycler.layoutManager = LinearLayoutManager(
                requireContext(), LinearLayoutManager.HORIZONTAL, false
        )

        familyViewModel.tweets.observe(viewLifecycleOwner, Observer { tweets ->
            when (tweets) {
                is Resource.Success -> twitterAdapter.refreshTweets(
                        tweets.data
                                .filter { it.authorPhone == phone }
                                .toArrayList()
                )
                is Resource.Loading -> twitterAdapter.refreshTweets(ArrayList())
                is Resource.Failure -> popBackStack()
            }
        })
    }

    private fun initFamilyView(userPhoneNumber: String) {
        familyViewModel.familyMembers.observe(viewLifecycleOwner, Observer {
            when (it) {
                is Resource.Success -> {
                    val memberResource = familyViewModel.getFamilyMemberByNumber(userPhoneNumber)
                    if (memberResource is Resource.Success) {
                        setCurrentData(FamilyMemberDto(memberResource.data))
                        setManagerMode(memberResource.data.fullPhoneNumber == getUserPhone())
                    } else {
                        popBackStack()
                    }
                }
                is Resource.Failure -> popBackStack()

            }
        })
    }

    private fun setCurrentData(dto: FamilyMemberDto) {
        binding.familyMemberViewNameText.text = dto.name
        if (dto.imageAddress != "") {
            binding.familyViewPhoto.setPadding(0, 0, 0, 0)
            Glide.with(requireContext()).load(dto.imageAddress).into(binding.familyViewPhoto)
        } else {
            val dpForPadding = getDp(requireContext(), 20)
            binding.familyViewPhoto.setPadding(dpForPadding, dpForPadding, dpForPadding, dpForPadding)
        }
        if (dto.birthDate == "") {
            binding.familyMemberViewBirthdateText.text = resources
                    .getString(R.string.family_member_view_date_of_birth_not_known)
            binding.familyMemberViewZodiacSign.visibility = View.GONE
        } else {
            binding.familyMemberViewBirthdateText.text = dto.birthDate
            binding.familyMemberViewZodiacSign.text = dto.zodiacSign
            binding.familyMemberViewZodiacSign.visibility = View.VISIBLE
        }
        if (dto.studyPlace == "") {
            binding.familyMemberViewStudyPlace.visibility = View.GONE
        } else {
            binding.familyMemberViewStudyPlace.visibility = View.VISIBLE
            binding.familyMemberViewStudyPlace.text = dto.studyPlace
        }
        if (dto.workPlace == "") {
            binding.familyMemberViewWorkPlace.visibility = View.GONE
        } else {
            binding.familyMemberViewWorkPlace.visibility = View.VISIBLE
            binding.familyMemberViewWorkPlace.text = dto.workPlace
        }

        if (dto.isOnline) {
            binding.familyMemberViewOnlineText.text = requireContext().getString(R.string.online_text);
            binding.familyMemberViewOnlineText.setBackgroundColor(resources.getColor(R.color.colorOnlineGreen));
        } else {
            binding.familyMemberViewOnlineText.text = requireContext().getString(R.string.offline_text);
            binding.familyMemberViewOnlineText.setBackgroundColor(resources.getColor(R.color.colorGray));
        }
    }

    private fun setManagerMode(managerMode: Boolean) {
        if (managerMode) {
            binding.familyMemberViewSettings.visibility = View.VISIBLE
            binding.familyMemberViewSettings.isEnabled = true
            binding.familyMemberViewSettings.setOnClickListener { v: View -> showPopUpSettingsMenu(v) }
            binding.familyMemberViewAddTweet.visibility = View.VISIBLE
        } else {
            binding.familyMemberViewSettings.visibility = View.INVISIBLE
            binding.familyMemberViewAddTweet.visibility = View.INVISIBLE
            binding.familyMemberViewSettings.isEnabled = false
        }
    }

    private fun showPopUpSettingsMenu(v: View) {
        val popupMenu = PopupMenu(context, v)
        val menuInflater = popupMenu.menuInflater
        menuInflater.inflate(R.menu.settings_person_page_menu, popupMenu.menu)
        popupMenu.setOnMenuItemClickListener { item: MenuItem ->
            when (item.itemId) {
                R.id.edit_person_page -> {
                    NavHostFragment.findNavController(this)
                            .navigate(R.id.memberPersonEditFragment)
                    return@setOnMenuItemClickListener true
                }
                else -> return@setOnMenuItemClickListener false
            }
        }
        popupMenu.show()
    }

    override fun onDestroy() {
        super.onDestroy()
        _binding = null
    }
}
